dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+#define DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+
+#include <utility>
+#include <type_traits>
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/span.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+/**
+ * @brief Calculates the position of most significant bit in x
+ * @return The position of MSB or 0 if x is 0.
+ */
+template <typename T>
+constexpr enable_if_t<std::is_integral<T>::value, int> ilog2(T x) {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+}
+
+/**
+ * @brief Returns an integer where bits at indicies in `bit_indices` are set to 1.
+ * @remarks Indices that are outside the bit-width of OutType are ignored.
+ */
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
+  OutType mask = 0;
+  for (int idx : bit_indices)
+    mask |= OutType(1) << idx;
+
+  return mask;
+}
+
+/**
+ * @brief Calculates equivalent, simpler reduction by collapsing adjacent dimensions.
+ *
+ * @param out_axes   a collection of integers, supporting `clear` and `push_back`.
+ * @param dim_groups a collection of integer pairs, supporting `clear` and `push_back`;
+ *                   the returned values are { index, count }, where index is the index of the
+ *                   first dimension in group and count is the number of collapsed dimensions.
+ *
+ * @param in_shape   shape of the input
+ * @param axes       list of reduced axes; axis indices must be <= 63
+ *
+ * A dimension can be collapsed with its neighbor if:
+ *  - its extent is 1 in all samples
+ *  - both dimensions are reduced or non-reduced.
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
+                       const TensorListShape<ndim> &in_shape, span<const int> axes) {
+  out_axes.clear();
+  dim_groups.clear();
+  int d = in_shape.sample_dim();
+
+  uint64_t mask = to_bit_mask(axes);
+
+  int i;
+
+  // skip leading degenerate dimensions
+  for (i = 0; i < d && is_degenerate_dim(in_shape, i); i++) {}
+
+  if (i < d) {
```

Comment:
Can you maybe reverse the condition, do early exit in if case, and do the rest without else?

Reply:
Why not.